backend/src/process.py

```diff
@@ -109,34 +110,52 @@ def __create_iterator_executor(self) -> Executor:
             parent_executor=self.executor,
         )
 
-    async def run_iteration(self, index: int, total: int):
-        executor = self.__create_iterator_executor()
+    def __get_eta(self, index: int, total: int) -> float:
+        if len(self.times) == 0:
+            return 0
+        return (sum(self.times) / len(self.times)) * (total - index)
 
-        await self.progress.suspend()
+    async def __update_progress(self, index: int, length: int):
         await self.executor.queue.put(
             {
                 "event": "iterator-progress-update",
                 "data": {
-                    "percent": index / total,
+                    "percent": index / length,
+                    "index": index,
+                    "total": length,
+                    "eta": self.__get_eta(index, length),
                     "iteratorId": self.iterator_id,
                     "running": list(self.chain.nodes.keys()),
                 },
             }
         )
 
-        try:
-            await executor.run_iteration(self.chain)
-        finally:
-            await self.executor.queue.put(
-                {
-                    "event": "iterator-progress-update",
-                    "data": {
-                        "percent": (index + 1) / total,
-                        "iteratorId": self.iterator_id,
-                        "running": None,
-                    },
-                }
-            )
+    async def __finish_progress(self, length: int):
+        await self.executor.queue.put(
+            {
+                "event": "iterator-progress-update",
+                "data": {
+                    "percent": 1,
+                    "index": length + 1,
+                    "total": length + 1,
+                    "eta": 0,
+                    "iteratorId": self.iterator_id,
+                    "running": None,
+                },
+            }
+        )
+
+    async def run_iteration(self, index: int, total: int):
+        executor = self.__create_iterator_executor()
+
+        await self.progress.suspend()
+
+        start = time.time()
+        await executor.run_iteration(self.chain)
+        end = time.time()
+        self.times.append(end - start)
+
+        await self.__update_progress(index + 1, total)
```

Comment:
`executor.run_iteration` might throw an error, but we still have to update progress. So I would suggest doing progress updates before calling the executor. This has a few other advantages as well, as I outlined in my previous comment.

```suggestion
        await self.progress.suspend()
        await self.__update_progress(index, total)

        start = time.time()
        await executor.run_iteration(self.chain)
        end = time.time()
        self.times.append(end - start)
```

Reply:
I tried that, and it was getting stuck at length - 1. Putting it after fixed it. Can't really explain that